pkg/spoke/controllers/statuscontroller/availablestatus_controller.go

```diff
@@ -0,0 +1,265 @@
+// ControllerSyncInterval is exposed so that integration tests can crank up the constroller resync speed.
+var ControllerReSyncInterval = 30 * time.Second
+
+// AvailableStatusController is to update the available status conditions of both manifests and manifestworks.
+type AvailableStatusController struct {
+	manifestWorkClient workv1client.ManifestWorkInterface
+	manifestWorkLister worklister.ManifestWorkNamespaceLister
+	spokeDynamicClient dynamic.Interface
+}
+
+// NewAvailableStatusController returns a AvailableStatusController
+func NewAvailableStatusController(
+	recorder events.Recorder,
+	spokeDynamicClient dynamic.Interface,
+	manifestWorkClient workv1client.ManifestWorkInterface,
+	manifestWorkInformer workinformer.ManifestWorkInformer,
+	manifestWorkLister worklister.ManifestWorkNamespaceLister,
+) factory.Controller {
+	controller := &AvailableStatusController{
+		manifestWorkClient: manifestWorkClient,
+		manifestWorkLister: manifestWorkLister,
+		spokeDynamicClient: spokeDynamicClient,
+	}
+
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, _ := meta.Accessor(obj)
+			return accessor.GetName()
+		}, manifestWorkInformer.Informer()).
+		WithSync(controller.sync).ResyncEvery(ControllerReSyncInterval).ToController("AvailableStatusController", recorder)
+}
+
+func (c *AvailableStatusController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	manifestWorkName := controllerContext.QueueKey()
+	if manifestWorkName != "key" {
+		// sync a particular manifestwork
+		manifestWork, err := c.manifestWorkLister.Get(manifestWorkName)
+		if errors.IsNotFound(err) {
+			// work not found, could have been deleted, do nothing.
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("unable to fetch manifestwork %q: %w", manifestWorkName, err)
+		}
+
+		err = c.syncManifestWork(ctx, manifestWork)
+		if err != nil {
+			return fmt.Errorf("unable to sync manifestwork %q: %w", manifestWork.Name, err)
+		}
+		return nil
+	}
+
+	// resync all manifestworks
+	klog.V(4).Infof("Reconciling all ManifestWorks")
+	manifestWorks, err := c.manifestWorkLister.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("unable to list manifestworks: %w", err)
+	}
+
+	errs := []error{}
+	for _, manifestWork := range manifestWorks {
+		err = c.syncManifestWork(ctx, manifestWork)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("unable to sync manifestwork %q: %w", manifestWork.Name, err))
+		}
+	}
+	if len(errs) > 0 {
+		return fmt.Errorf("unable to resync manifestworks: %w", utilerrors.NewAggregate(errs))
+	}
+	return nil
+}
+
+func (c *AvailableStatusController) syncManifestWork(ctx context.Context, originalManifestWork *workapiv1.ManifestWork) error {
+	klog.V(4).Infof("Reconciling ManifestWork %q", originalManifestWork.Name)
+	manifestWork := originalManifestWork.DeepCopy()
+
+	// no work if the manifestwork is terminating
+	if !manifestWork.DeletionTimestamp.IsZero() {
```

Comment:
I think we should still update available status when it is being deleted.

Reply:
Fixed